File: go/internal/store/bookings.go
package store

import (
	"context"
	"fmt"

	"github.com/mudbase/mudbase-showcase-events/go/internal/mbase"
	"github.com/mudbase/mudbase-showcase-events/go/internal/models"
)

// myBookingsLimit mirrors the reference web app's useMyBookings limit
// (../web/src/hooks/useBookings.ts).
const myBookingsLimit = 100

// BookingService implements every operation the app needs against the `bookings` collection.
type BookingService struct {
	client       *mbase.Client
	collectionID string
	activity     *ActivityService
}

// NewBookingService builds a BookingService bound to the given bookings collection ID.
func NewBookingService(client *mbase.Client, bookingsCollectionID string, activity *ActivityService) *BookingService {
	return &BookingService{client: client, collectionID: bookingsCollectionID, activity: activity}
}

// ByID fetches a single booking by its Mudbase document ID.
func (s *BookingService) ByID(ctx context.Context, token, id string) (models.Booking, error) {
	booking, err := mbase.Get[models.Booking](ctx, s.client, token, s.collectionID, id)
	if err != nil {
		return models.Booking{}, fmt.Errorf("store: fetching booking %s: %w", id, err)
	}
	return booking, nil
}

// MyBookings returns the signed-in attendee's own bookings across every event, newest first.
func (s *BookingService) MyBookings(ctx context.Context, token, userID string) ([]models.Booking, error) {
	result, err := mbase.List[models.Booking](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"userId": userID},
		Sort:   "-createdAt",
		Limit:  myBookingsLimit,
	})
	if err != nil {
		return nil, fmt.Errorf("store: listing bookings for user %s: %w", userID, err)
	}
	return result.Data, nil
}

// MyBookingForEvent returns the signed-in user's own booking for one specific event, if any - used
// to hide the Book button / show its status instead.
func (s *BookingService) MyBookingForEvent(ctx context.Context, token, eventID, userID string) (models.Booking, bool, error) {
	result, err := mbase.List[models.Booking](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"eventId": eventID, "userId": userID},
		Limit:  1,
	})
	if err != nil {
		return models.Booking{}, false, fmt.Errorf("store: fetching booking for event %s / user %s: %w", eventID, userID, err)
	}
	if len(result.Data) == 0 {
		return models.Booking{}, false, nil
	}
	return result.Data[0], true, nil
}

// ConfirmedCount returns the live server-side confirmed-booking count for one event, for the
// capacity indicator on its card/detail page.
func (s *BookingService) ConfirmedCount(ctx context.Context, token, eventID string) (int, error) {
	result, err := mbase.List[models.Booking](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"eventId": eventID, "status": string(models.StatusConfirmed)},
		Limit:  1,
	})
	if err != nil {
		return 0, fmt.Errorf("store: counting confirmed bookings for event %s: %w", eventID, err)
	}
	return int(result.Total), nil
}

// ByEvent returns every booking for one event (used by the organizer-only check-in lookup and by
// ConfirmedCount's sibling reconciliation pass).
func (s *BookingService) ByEvent(ctx context.Context, token, eventID string) ([]models.Booking, error) {
	result, err := mbase.List[models.Booking](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"eventId": eventID},
		Sort:   "createdAt",
		Limit:  reconcileFetchLimit,
	})
	if err != nil {
		return nil, fmt.Errorf("store: listing bookings for event %s: %w", eventID, err)
	}
	return result.Data, nil
}

// Create books eventID for the given attendee, using the capacity-race approach documented in
// ../../web/plan/build-plan.md: decide the initial status from a fresh server-side confirmed
// count, write it, then run the shared reconciliation pass so a race against another concurrent
// booking self-corrects. Returns the booking's *post-reconciliation* state, not its tentative
// initial write, so the caller never reports a status that got corrected out from under it a
// moment later. Ported from the reference web app's useCreateBooking.
func (s *BookingService) Create(ctx context.Context, token, eventID string, capacity int, userID, userName string) (models.Booking, error) {
	confirmedCount, err := s.ConfirmedCount(ctx, token, eventID)
	if err != nil {
		return models.Booking{}, err
	}
	initialStatus := models.StatusWaitlisted
	if confirmedCount < capacity {
		initialStatus = models.StatusConfirmed
	}

	qrToken, err := generateQRToken()
	if err != nil {
		return models.Booking{}, err
	}

	booking, err := mbase.Create[models.Booking](ctx, s.client, token, s.collectionID, map[string]interface{}{
		"eventId":  eventID,
		"userId":   userID,
		"userName": userName,
		"status":   string(initialStatus),
		"qrToken":  qrToken,
	})
	if err != nil {
		return models.Booking{}, fmt.Errorf("store: creating booking for event %s: %w", eventID, err)
	}

	confirmedAction := models.ActionBookingWaitlisted
	if initialStatus == models.StatusConfirmed {
		confirmedAction = models.ActionBookingConfirmed
	}
	if err := s.activity.Log(ctx, token, eventID, userID, userName, confirmedAction); err != nil {
		return models.Booking{}, err
	}

	if err := ReconcileEventCapacity(ctx, s.client, s.collectionID, s.activity, token, eventID, capacity); err != nil {
		return models.Booking{}, err
	}

	// Re-read: reconciliation above may have demoted this exact booking if it lost a race against
	// another concurrent request that also decided "confirmed" from the same pre-write count.
	final, err := mbase.Get[models.Booking](ctx, s.client, token, s.collectionID, booking.ID)
	if err != nil {
		return models.Booking{}, fmt.Errorf("store: re-fetching booking %s after reconciliation: %w", booking.ID, err)
	}
	return final, nil
}

// Cancel cancels an attendee's own booking, then reconciles so the earliest waitlisted booking is
// promoted into the freed seat. Ported from the reference web app's useCancelBooking.
func (s *BookingService) Cancel(ctx context.Context, token, bookingID, eventID string, capacity int, userID, userName string) error {
	if _, err := mbase.Update[models.Booking](ctx, s.client, token, s.collectionID, bookingID, map[string]interface{}{
		"status": string(models.StatusCancelled),
	}); err != nil {
		return fmt.Errorf("store: cancelling booking %s: %w", bookingID, err)
	}

	if err := s.activity.Log(ctx, token, eventID, userID, userName, models.ActionBookingCancelled); err != nil {
		return err
	}

	return ReconcileEventCapacity(ctx, s.client, s.collectionID, s.activity, token, eventID, capacity)
}

// CheckInOutcome enumerates every result CheckIn can produce.
type CheckInOutcome string

const (
	CheckInOutcomeCheckedIn      CheckInOutcome = "checked_in"
	CheckInOutcomeAlreadyChecked CheckInOutcome = "already_checked_in"
	CheckInOutcomeCancelled      CheckInOutcome = "cancelled"
	CheckInOutcomeWaitlisted     CheckInOutcome = "waitlisted"
	CheckInOutcomeNotFound       CheckInOutcome = "not_found"
)

// CheckInResult bundles the outcome and (when found) the matched booking.
type CheckInResult struct {
	Outcome CheckInOutcome
	Booking models.Booking
}

// CheckIn looks up a booking by its scanned/pasted qrToken within one event and, if eligible,
// checks it in. Ported from the reference web app's useCheckIn.
func (s *BookingService) CheckIn(ctx context.Context, token, eventID, qrToken string) (CheckInResult, error) {
	if qrToken == "" {
		return CheckInResult{Outcome: CheckInOutcomeNotFound}, nil
	}

	result, err := mbase.List[models.Booking](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"eventId": eventID, "qrToken": qrToken},
		Limit:  1,
	})
	if err != nil {
		return CheckInResult{}, fmt.Errorf("store: looking up qr token for event %s: %w", eventID, err)
	}
	if len(result.Data) == 0 {
		return CheckInResult{Outcome: CheckInOutcomeNotFound}, nil
	}
	booking := result.Data[0]

	switch booking.Status {
	case models.StatusCheckedIn:
		return CheckInResult{Outcome: CheckInOutcomeAlreadyChecked, Booking: booking}, nil
	case models.StatusCancelled:
		return CheckInResult{Outcome: CheckInOutcomeCancelled, Booking: booking}, nil
	case models.StatusWaitlisted:
		return CheckInResult{Outcome: CheckInOutcomeWaitlisted, Booking: booking}, nil
	}

	updated, err := mbase.Update[models.Booking](ctx, s.client, token, s.collectionID, booking.ID, map[string]interface{}{
		"status": string(models.StatusCheckedIn),
	})
	if err != nil {
		return CheckInResult{}, fmt.Errorf("store: checking in booking %s: %w", booking.ID, err)
	}
	if err := s.activity.Log(ctx, token, eventID, booking.UserID, booking.UserName, models.ActionCheckedIn); err != nil {
		return CheckInResult{}, err
	}
	return CheckInResult{Outcome: CheckInOutcomeCheckedIn, Booking: updated}, nil
}

File: python/app/services/bookings.py
"""bookings collection: both roles create/read their own bookings (Mudbase
collection permissions scope reads/writes to `userId === session.user.id`
for the attendee role; organizer bookings aren't modeled as a separate write
path in this UI — see ../web/plan/build-plan.md's RBAC matrix). Mirrors
../web/src/hooks/useBookings.ts field-for-field, including the exact
capacity-decide-then-reconcile approach documented there and ported in
`app/services/capacity.py`.
"""

import asyncio

from app.config import get_settings
from app.mudbase_client import get_data_sync, list_data_sync, update_data_sync
from app.mudbase_client import create_data_sync as _create_data_sync
from app.schemas.booking import BookingDoc
from app.services.activity import log_activity
from app.services.capacity import reconcile_event_capacity
from app.utils import generate_qr_token

_MY_BOOKINGS_LIMIT = 100  # DataApi.list_data's server-enforced cap.

CheckInOutcome = str  # "checked_in" | "already_checked_in" | "cancelled" | "waitlisted" | "not_found"


async def list_my_bookings(user_id: str, *, access_token: str) -> list[BookingDoc]:
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.bookings_collection_id,
        filter_dict={"userId": user_id},
        sort="-createdAt",
        limit=_MY_BOOKINGS_LIMIT,
        access_token=access_token,
    )
    return [BookingDoc.model_validate(doc) for doc in result["data"]]


async def get_booking_for_event(event_id: str, user_id: str, *, access_token: str) -> BookingDoc | None:
    """The signed-in user's own booking for one specific event, if any (used
    to hide the Book button / show its status instead — mirrors
    ../web/src/hooks/useBookings.ts::useMyBookingForEvent)."""
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.bookings_collection_id,
        filter_dict={"eventId": event_id, "userId": user_id},
        limit=1,
        access_token=access_token,
    )
    docs = result["data"]
    return BookingDoc.model_validate(docs[0]) if docs else None


async def get_booking(booking_id: str, *, access_token: str) -> BookingDoc | None:
    settings = get_settings()
    doc = await asyncio.to_thread(get_data_sync, settings.bookings_collection_id, booking_id, access_token=access_token)
    return BookingDoc.model_validate(doc) if doc else None


async def get_confirmed_count(event_id: str, *, access_token: str) -> int:
    """The live confirmed-booking count for one event, for the capacity
    indicator on its card/detail — mirrors
    ../web/src/hooks/useEvents.ts::useConfirmedCount (a real server-side
    count via `pagination.total`, not a client-cached guess)."""
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.bookings_collection_id,
        filter_dict={"eventId": event_id, "status": "confirmed"},
        limit=1,
        access_token=access_token,
    )
    pagination = result["pagination"] or {}
    total: int = pagination.get("total", 0)
    return total


async def create_booking(
    event_id: str,
    capacity: int,
    *,
    user_id: str,
    user_name: str,
    access_token: str,
) -> BookingDoc:
    """Mirrors ../web/src/hooks/useBookings.ts::useCreateBooking exactly:
    decide the initial status from a fresh server-side confirmed count,
    write it, log activity, run the shared reconciliation pass, then
    re-read the booking's *post-reconciliation* state so the caller never
    reports a status that got corrected out from under it a moment later."""
    settings = get_settings()
    confirmed_count = await get_confirmed_count(event_id, access_token=access_token)
    initial_status = "confirmed" if confirmed_count < capacity else "waitlisted"
    qr_token = generate_qr_token()

    created_doc = await asyncio.to_thread(
        _create_data_sync,
        settings.bookings_collection_id,
        {"eventId": event_id, "userId": user_id, "userName": user_name, "status": initial_status, "qrToken": qr_token},
        access_token=access_token,
    )
    booking = BookingDoc.model_validate(created_doc)

    await log_activity(
        event_id=event_id,
        actor_id=user_id,
        actor_name=user_name,
        action="booking_confirmed" if initial_status == "confirmed" else "booking_waitlisted",
        access_token=access_token,
    )

    await reconcile_event_capacity(event_id, capacity, access_token=access_token)

    refreshed = await asyncio.to_thread(
        get_data_sync, settings.bookings_collection_id, booking.id, access_token=access_token
    )
    return BookingDoc.model_validate(refreshed) if refreshed else booking


async def cancel_booking(booking: BookingDoc, capacity: int, *, access_token: str) -> None:
    """Mirrors ../web/src/hooks/useBookings.ts::useCancelBooking: cancels,
    logs activity, then reconciles so the earliest waitlisted booking is
    promoted into the freed seat."""
    settings = get_settings()
    await asyncio.to_thread(
        update_data_sync, settings.bookings_collection_id, booking.id, {"status": "cancelled"}, access_token=access_token
    )
    await log_activity(
        event_id=booking.event_id,
        actor_id=booking.user_id,
        actor_name=booking.user_name,
        action="booking_cancelled",
        access_token=access_token,
    )
    await reconcile_event_capacity(booking.event_id, capacity, access_token=access_token)


async def check_in_by_qr_token(
    event_id: str, qr_token: str, *, access_token: str
) -> tuple[CheckInOutcome, BookingDoc | None]:
    """Mirrors ../web/src/hooks/useBookings.ts::useCheckIn's outcome ladder
    exactly: not found, already checked in (idempotent), cancelled,
    waitlisted (cannot check in until promoted), or a fresh check-in."""
    settings = get_settings()
    trimmed = qr_token.strip()
    if not trimmed:
        return "not_found", None

    result = await asyncio.to_thread(
        list_data_sync,
        settings.bookings_collection_id,
        filter_dict={"eventId": event_id, "qrToken": trimmed},
        limit=1,
        access_token=access_token,
    )
    docs = result["data"]
    if not docs:
        return "not_found", None

    booking = BookingDoc.model_validate(docs[0])
    if booking.status == "checked_in":
        return "already_checked_in", booking
    if booking.status == "cancelled":
        return "cancelled", booking
    if booking.status == "waitlisted":
        return "waitlisted", booking

    updated_doc = await asyncio.to_thread(
        update_data_sync, settings.bookings_collection_id, booking.id, {"status": "checked_in"}, access_token=access_token
    )
    updated = BookingDoc.model_validate(updated_doc)
    await log_activity(
        event_id=event_id,
        actor_id=booking.user_id,
        actor_name=booking.user_name,
        action="checked_in",
        access_token=access_token,
    )
    return "checked_in", updated
